core/planner/planner.go

```diff
@@ -32,61 +32,41 @@ func New(log *util.Logger, tariff api.Tariff, opt ...func(t *Planner)) *Planner
 }
 
 // plan creates a lowest-cost plan or required duration.
-// It MUST already established that
+// It MUST already be established that:
 // - rates are sorted in ascending order by cost and descending order by start time (prefer late slots)
-// - target time and required duration are before end of rates
+// - rates are filtered to [now, targetTime] window by caller
 func (t *Planner) plan(rates api.Rates, requiredDuration time.Duration, targetTime time.Time) api.Rates {
 	var plan api.Rates
 
-	for _, source := range rates {
-		// slot not relevant
-		if !(source.End.After(t.clock.Now()) && source.Start.Before(targetTime)) {
-			continue
-		}
-
-		// adjust slot start and end
-		slot := source
-		if slot.Start.Before(t.clock.Now()) {
-			slot.Start = t.clock.Now()
-		}
-		if slot.End.After(targetTime) {
-			slot.End = targetTime
+	for _, slot := range rates {
+		if requiredDuration <= 0 {
+			break
 		}
 
 		slotDuration := slot.End.Sub(slot.Start)
-		requiredDuration -= slotDuration
 
 		// slot covers more than we need, so shorten it
-		if requiredDuration < 0 {
-			// the first (if not single) slot should start as late as possible
+		if slotDuration > requiredDuration {
```

Comment:
I'm trying to revert this section to the original form:

```go
	for _, slot := range rates {
		slotDuration := slot.End.Sub(slot.Start)
		requiredDuration -= slotDuration

		// slot covers more than we need, so shorten it
		if requiredDuration < 0 {
			if IsFirst(slot, plan) && len(plan) > 0 {
				slot.Start = slot.Start.Add(slotDuration - requiredDuration)
			} else {
				slot.End = slot.End.Add(requiredDuration - slotDuration)
			}
			requiredDuration = 0
		}

		plan = append(plan, slot)

		// we found all necessary slots
		if requiredDuration == 0 {
			break
		}
	}
```

But tests keep failing. I don't get why- these two forms should be equivalant.